### PR TITLE
ml-mod.it

### PR DESCRIPTION
**Pull request by papadeiv**
Inference pipeline for machine-learning detection of toxicity and profanity content in italian and english comments.

### CODE REVIEW

File: ml/requirements.txt [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+urllib3==1.26.7
+w3lib==1.22.0
+wandb==0.12.9
+wcwidth @ file:///Users/ktietz/demo/mc3/conda-bld/wcwidth_1629357192024/work
+webencodings==0.5.1
+websocket-client==1.2.3
+websockets==10.1
+widgetsnbextension==3.5.1
+yaspin==2.1.0
+zipp @ file:///tmp/build/80754af9/zipp_1633618647012/work
```
Comment (e-caste) [inline ml/requirements.txt:153]: Requirements should only be the needed libraries to make everything work (including the scrapers), and they shouldn't contain any local file. If `conda` uses the strange `library @ file:///path` format then simply use `python3 -m venv venv` followed by `source venv/bin/activate` and `pip install <libraries>` and finally `pip freeze > requirements.txt` to get a clean requirements file.

File: ml/Detectors.py [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+
+  def __init__(self, language_id):
+    if language_id == 0:
+      # Italian dictionary of swear/stop-words is built
+      self.stopwords = ["per", "è", "é", "e", "il", "che", "tua", "sua", "un", "uno", "una"]
+      self.vocabulary = ["merda", "cazzo", "incazzato", "stocazzo", "coglione", "minchia", "minchiata", "minchiate", "stronzo", "stronzata", "stronzate", "puttana", "puttanata", "puttanate", "vaffanculo", "culo", "ricchione", "frocio", "incel"]
+    else:
+      # English dictionary of swear/stop-words is built
+      self.stopwords = ["per", "è", "é", "e", "il", "che", "tua", "sua", "un", "uno", "una"]
+      self.vocabulary = ["merda", "cazzo", "incazzato", "stocazzo", "coglione", "minchia", "minchiata", "minchiate", "stronzo", "stronzata", "stronzate", "puttana", "puttanata", "puttanate", "vaffanculo", "culo", "ricchione", "frocio", "incel"]
```
Comment (e-caste) [inline ml/Detectors.py:23]: See issue #91

File: ml/Detectors.py [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+    self.hyperparams = TrainingArguments(output_dir='./train_results',
+                                         num_train_epochs=3,
+                                         per_device_train_batch_size=8,
+                                         per_device_eval_batch_size=16,
+                                         warmup_steps=500,
+                                         weight_decay=0.01,
+                                         logging_dir='./train_logs',
+                                         logging_steps=10,
+                                         report_to='wandb'
+                                         )
```
Comment (e-caste) [inline ml/Detectors.py:83]: Could you add a bit of documentation (either in the code or in the README) as to where to find the correct hyperparameters for each model?

File: ml/Detectors.py [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+
+
+  def banning_policy(self, state_distribution):
+    # Extract toxicity states from labels' probability distribution
+    non_toxic, toxic = state_distribution[0,0], state_distribution[0,1]
+    # Define and enforce the bannong policy
+    if toxic > 0.5:
+      return True  # Toxic
+    else:
+      return False # Non-Toxic
```
Comment (e-caste) [inline ml/Detectors.py:135]:
```suggestion
    return toxic > 0.5
```

File: ml/Modali.py [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+        if language_id < 13:
+          if filtered_argmax[0][0] == 12:
+            return 13
+          else:
+            return None
+        else:
+          if filtered_argmax[0][0] == 13:
+            return 13
+          else:
+            return None
```
Comment (e-caste) [inline ml/Modali.py:89]: See issue #92

File: ml/Modali.py [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+  train()
+
+  # Filter unwanted characters out of the input text
+  sample = comment.replace("[", "").replace("]", "").replace('\\n', ' ').replace('\\', ' ').replace("'", ""). replace('"', '')
+
+  # Initialise score for swear-words reference (True=ban, False=pass)
+  score = True
+
+  # Discern language of the comment between english and italian...
+  Linguist = Polyglot()
```
Comment (e-caste) [inline ml/Modali.py:119]: Just for the sake of clarity, variables should have a lowercase initial (while classes an uppercase one)
```suggestion
  linguist = Polyglot()
```

File: ml/test.py [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+non_toxic = 0
+
+for j in range(0, len(features)):
+    if labels[j] == True: # Toxic
+        dataset.append({"text": str(features[j]), "label": toxic_id})
+        toxic += 1
+
+    elif labels[j] == False: # Non-Toxic
+        dataset.append({"text": str(features[j]), "label": non_toxic_id})
+        non_toxic += 1
```
Comment (e-caste) [inline ml/test.py:46]: Just a bit more pythonic:
```suggestion
for label, feature in zip(labels, features):
    if label is True:  # Toxic
        dataset.append({"text": str(feature), "label": toxic_id})
        toxic += 1

    elif label is False:  # Non-Toxic
        dataset.append({"text": str(feature), "label": non_toxic_id})
        non_toxic += 1
```

File: ml/test.py [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+for j in range(cardinality):
+    comment = str(new_features[j])
+    sample = comment.replace("[", "").replace("]", "").replace('\\n', ' ').replace('\\', ' ').replace("'", ""). replace('"', '')
+    label = infer(sample)
+    
+    # Redirect model's inferred label (True/False) to numeric values (1/0)
+    if label == True:
+        model_label = 1 # Toxic
+    else:
+        model_label = 0 # Non-Toxic
```
Comment (e-caste) [inline ml/test.py:76]:
```suggestion
    model_label = 1 if label is True else 0
```

**Review by e-caste — CHANGES_REQUESTED** [review]
Just a few notes until issues #91 and #92 are closed. Good job!

File: ml/test.py (hunk 7 above)
Comment (papadeiv) [reply ml/test.py:46]: Generates empty list.

File: ml/Modali.py (hunk 6 above)
Comment (papadeiv) [reply ml/Modali.py]: Solved with [357c76b](https://github.com/WEEE-Open/skeeelled/commit/357c76b7933afd7f3243c3b1b7b8146b14bd2524).

File: ml/Detectors.py (hunk 3 above)
Comment (papadeiv) [reply ml/Detectors.py:84]: Issue closed.

**Review by e-caste — APPROVED** [review]

**Pull request merged**